clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -431,6 +444,83 @@ private static int parseAcks(String acksString) {
     }
 
     /**
+     * Needs to be called before any other methods when the transactional.id is set in the configuration.
+     *
+     * This method does the following:
+     *   1. Ensures any transactions initiated by previous instances of the producer
+     *      are completed. If the previous instance had failed with a transaction in
+     *      progress, it will be aborted. If the last transaction had begun completion,
+     *      but not yet finished, this method awaits its completion.
+     *   2. Gets the internal producer id and epoch, used in all future transactional
+     *      messages issued by the producer.
+     *
+     * @throws IllegalStateException if the TransactionalId for the producer is not set
+     *         in the configuration.
+     */
+    public void initTransactions() {
+        if (transactionState == null || !transactionState.isTransactional()) {
+            throw new IllegalStateException("Cannot call initTransactions without setting a transactional id.");
+        }
+        FutureTransactionalResult result = transactionState.initializeTransactions();
+        result.get();
+    }
+
+    /**
+     * Should be called before the start of each new transaction.
+     *
+     * @throws ProducerFencedException if another producer is with the same
+     *         transactional.id is active.
+     */
+    public void beginTransaction() throws ProducerFencedException {
+        // Set the transactional bit in the producer.
+        assert transactionState != null && transactionState.isTransactional() && !transactionState.isInTransaction() && transactionState.hasPid();
```

Comment:
We don't usually use JVM level asserts because they are disabled by default. Same for all other cases in this PR.